Remove extraneous code from retrievers

jdconrad opened the pull request:
This change removes all of the serialization code from retrievers since these will only be used as x-content parsers extracting values into the `SearchSourceBuilder`. This drastically reduces the test coverage area required. This change also removes post filter from the standard retriever for now as a post filter may need to apply to all returned documents in a compound query. Finally, this adds some additional checks for values that should not be shared between global values and retriever values.

jdconrad committed:
remove unncessary code

elasticsearchmachine commented:
Pinging @elastic/es-search (Team:Search)

mayya-sharipova commented on server/src/main/java/org/elasticsearch/search/retriever/RetrieverBuilder.java:
```diff
@@ -162,40 +151,13 @@ protected static RetrieverBuilder<?> parseInnerRetrieverBuilder(XContentParser p
-    public abstract void doWriteTo(StreamOutput out) throws IOException;
 
     @Override
     public final XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
```
Do we need `toXContent` method at all? When will it be used?

mayya-sharipova reviewed (approved):
@jdconrad Thanks.

Just to clarify, a coordinating node upon receiving a search request with retrievers converts it to a standard search request, and all subsequent interactions with this request (e.g. passing it to shards) will have no info about retrievers. Is this right?

This simplification makes sense to me. I've added a comment if we can also remove `toXContent` method, as I am not sure how it is used.

jdconrad commented:
@mayya-sharipova Thank you for the review! I will commit as soon as I update this with your suggested changes.

You are correct that these should only ever be used to parse a search request containing retrievers into a standard search request. I removed toXContent as it is not used.

I also removed linear_combination retriever for now as it's behavior is still open to discussion.

jdconrad committed:
additional removal
